glean-core/csharp/Glean/Dispatchers.cs

```diff
@@ -3,24 +3,213 @@
+        /// immediately and synchronously to avoid asynchronous issues in tests.
+        /// </summary>
+        /// <param name="action">The Action to invoke</param>
         internal static void LaunchAPI(Action action)
         {
-            action();
+            if (QueueInitialTasks)
+            {
+                // If we are queuing tasks, typically before Glean has been
+                // initialized, then we should just add the created Task to the
+                // taskQueue.
+                AddTaskToQueue(new Task(action, cancellationToken));
+            }
+            else
+            {
+                if (!TestingMode)
+                {
+                    // If we are not queuing initial tasks, we can go ahead and
+                    // execute the task asynchronously on the ExclusiveScheduler
+                    new Task(action, cancellationToken)
+                        .Start(taskSchedulerPair.ExclusiveScheduler);
+                }
+                else
+                {
+                    // If we are in testing mode, then go ahead and invoke the
+                    // action to ensure synchronous execution.
+                    action.Invoke();
+                }
+            }
+         
         }
 
-        internal static void FlushQueuedInitialTasks()
+        /// <summary>
+        /// This function launches an Action as an asynchronous Task on a
+        /// concurrently executed queue.
+        ///
+        /// **Note:** Tasks that need to be executed asynchronously but should
+        /// not block other tasks such as recording data should use the
+        /// `LaunchConcurrent` function. For example, tasks performed during
+        /// initialization or an upload task could be executed concurrently.
+        /// For tasks that need to be executed serially, see `LaunchAPI`.
+        ///
+        /// This function specifically ignores the `QueueInitialTasks` flag
+        /// because the only tasks that should be launched by this are the ping
+        /// upload schedulers and those should run regardless of the initialized
+        /// state.
+        ///
+        /// If `TestingMode` is enabled, then `LaunchConcurrent` will just
+        /// execute the task rather than adding it to the concurrent queue to
+        /// avoid asynchronous issues while testing.
+        /// </summary>
+        /// <param name="action">The Action to perform</param>
+        internal static void LaunchConcurrent(Action action)
         {
+            if (TestingMode)
+            {
+                action.Invoke();        
+            }
+            else
+            {
+                new Task(action, cancellationToken)
+                    .Start(taskSchedulerPair.ConcurrentScheduler);
+            }
+        }
+
+        /// <summary>
+        /// Cancels any pending background tasks.
+        /// </summary>
+        internal static void CancelBackgroundTasks()
```

Comment:
What's this for?

Reply:
We cancel background tasks when we toggle upload enabled to false, this is just a way to cancel all of the tasks running in the Glean background queues.  This is similar to what we are doing when we [cancel the `PingUploadWorker` in Kotlin](https://github.com/mozilla/glean/blob/5ecdd24b1dc92539903bf2246cdc0aa3a56954d9/glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt#L274) and [cancelling the background tasks in Swift](https://github.com/mozilla/glean/blob/5ecdd24b1dc92539903bf2246cdc0aa3a56954d9/glean-core/ios/Glean/Glean.swift#L222).  In Kotlin, there just wasn't a good way to cancel a coroutine, if I remember correctly.